### PR TITLE
Specify IBufferByteAccess as a RuntimeType

### LINKED ISSUE #346: InvalidCastException on StreamReader.ReadToEndAsync
This code:
```cs
Windows.Storage.StorageFile file = ...;
string text = "";
using (var reader = new System.IO.StreamReader(await file.OpenStreamForReadAsync(), true))
{
    text = await reader.ReadToEndAsync();
}
```
throws at the ReadToEndAsync, with this stack:
```
 	winrt.runtime.dll!WinRT.ExceptionHelpers.ThrowExceptionForHR(int hr) Line 69	C#
>	Microsoft.Windows.SDK.NET.dll!ABI.Windows.Storage.Streams.IInputStream.ReadAsync(Windows.Storage.Streams.IBuffer buffer, uint count, Windows.Storage.Streams.InputStreamOptions options) Line 4505	C#
 	Microsoft.Windows.SDK.NET.dll!Windows.Storage.Streams.FileRandomAccessStream.ReadAsync(Windows.Storage.Streams.IBuffer buffer, uint count, Windows.Storage.Streams.InputStreamOptions options) Line 926	C#
 	Microsoft.Windows.SDK.NET.dll!Windows.Storage.Streams.FileRandomAccessStream.Windows.Storage.Streams.IInputStream.ReadAsync(Windows.Storage.Streams.IBuffer buffer, uint count, Windows.Storage.Streams.InputStreamOptions options) Line 928	C#
 	Microsoft.Windows.SDK.NET.dll!Windows.Storage.Streams.WinRtToNetFxStreamAdapter.ReadAsyncInternal(byte[] buffer, int offset, int count, System.Threading.CancellationToken cancellationToken) Line 9104	C#
 	Microsoft.Windows.SDK.NET.dll!Windows.Storage.Streams.WinRtToNetFxStreamAdapter.ReadAsync(byte[] buffer, int offset, int count, System.Threading.CancellationToken cancellationToken) Line 8855	C#
 	System.Private.CoreLib.dll!System.IO.Stream.ReadAsync(System.Memory<byte> buffer, System.Threading.CancellationToken cancellationToken) Line 455	C#
 	System.Private.CoreLib.dll!System.IO.BufferedStream.ReadFromUnderlyingStreamAsync(System.Memory<byte> buffer, System.Threading.CancellationToken cancellationToken, int bytesAlreadySatisfied, System.Threading.Tasks.Task semaphoreLockTask) Line 762	C#
 	System.Private.CoreLib.dll!System.IO.BufferedStream.ReadAsync(System.Memory<byte> buffer, System.Threading.CancellationToken cancellationToken) Line 711	C#
 	System.Private.CoreLib.dll!System.IO.StreamReader.ReadBufferAsync(System.Threading.CancellationToken cancellationToken) Line 1238	C#
 	System.Private.CoreLib.dll!System.IO.StreamReader.ReadToEndAsyncInternal() Line 931	C#
 	System.Private.CoreLib.dll!System.IO.StreamReader.ReadToEndAsync() Line 914	C#
```
aka
```
   at WinRT.ExceptionHelpers.ThrowExceptionForHR(Int32 hr) in d:\a\1\s\WinRT.Runtime\ExceptionHelpers.cs:line 69
   at ABI.Windows.Storage.Streams.IInputStream.ReadAsync(IBuffer buffer, UInt32 count, InputStreamOptions options) in C:\BA\232\s\NetSDKSln\obj\Release\net5.0\Generated Files\Windows.Storage.Streams.cs:line 4505
   at Windows.Storage.Streams.FileRandomAccessStream.ReadAsync(IBuffer buffer, UInt32 count, InputStreamOptions options) in C:\BA\232\s\NetSDKSln\obj\Release\net5.0\Generated Files\Windows.Storage.Streams.cs:line 926
   at Windows.Storage.Streams.FileRandomAccessStream.Windows.Storage.Streams.IInputStream.ReadAsync(IBuffer buffer, UInt32 count, InputStreamOptions options) in C:\BA\232\s\NetSDKSln\obj\Release\net5.0\Generated Files\Windows.Storage.Streams.cs:line 928
   at Windows.Storage.Streams.WinRtToNetFxStreamAdapter.<ReadAsyncInternal>d__43.MoveNext() in C:\BA\232\s\NetSDKSln\obj\Release\net5.0\Generated Files\Windows.Storage.Streams.cs:line 9104
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:line 57
   at Windows.Storage.Streams.WinRtToNetFxStreamAdapter.<ReadAsyncInternal>d__43.MoveNext() in C:\BA\232\s\NetSDKSln\obj\Release\net5.0\Generated Files\Windows.Storage.Streams.cs:line 9128
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:line 57
```

### PR DESCRIPTION
**Pull request by j0shuams**
CsWinRT was throwing an InvalidCastException when trying to read from a file. This PR provides marshaling support for the IBufferByteAccess interface, which previously was unsupported and causing a NoInterface error to be thrown. 

A unit test has been added for this scenario -- derived from a reproduction of an issue that was filed. This PR should close #346

### CODE REVIEW

**Comment by huoyaoyuan** [comment]
I wrote some test for this:
```csharp
using System;
using System.IO;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading.Tasks;
using Windows.Storage.Streams;
using Xunit;

namespace UnitTest
{
    public class StreamTests
    {
        [Fact]
        public async Task WriteBuffer()
        {
            var random = new Random(42);
            byte[] data = new byte[256];
            random.NextBytes(data);

            using var stream = new InMemoryRandomAccessStream();
            IBuffer buffer = data.AsBuffer();
            await stream.WriteAsync(buffer);
        }

        [Fact]
        public async Task WriteAndRead()
        {
            var random = new Random(42);
            byte[] data = new byte[256];
            random.NextBytes(data);

            using var stream = new InMemoryRandomAccessStream().AsStream();
            await stream.WriteAsync(data, 0, data.Length);
            stream.Seek(0, SeekOrigin.Begin);

            byte[] read = new byte[256];
            await stream.ReadAsync(read, 0, read.Length);
            Assert.Equal(read, data);
        }
    }
}
```

Could you confirm that these tests also get fixed?

**Comment by huoyaoyuan** [comment]
#359 should be the same issue

**Review by Scottj1s — APPROVED** [review]
thanks!

File: UnitTest/TestComponentCSharp_Tests.net5.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.IO;
+using TestComponentCSharp;
+using Windows.Storage;
+using Xunit;
+
+namespace UnitTest
+{
+    public class TestCSharpNet5
```
Comment (manodasanW) [inline UnitTest/TestComponentCSharp_Tests.net5.cs:9]: Any particular reason for putting this test in a separate file and the net5 suffx?  Anything net5 specific here?

File: UnitTest/TestComponentCSharp_Tests.net5.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            Windows.Storage.StorageFolder localFolder = Windows.Storage.ApplicationData.Current.LocalFolder;
+            string fileName = "IBufferByteAccessTest.txt";
+            Windows.Storage.StorageFile file = await localFolder.CreateFileAsync(fileName, CreationCollisionOption.ReplaceExisting);
+            string text = "";
+            using (var reader = new System.IO.StreamReader(await file.OpenStreamForReadAsync(), true))
+            {
+                try
+                {
+                    text = await reader.ReadToEndAsync();
```
Comment (manodasanW) [inline UnitTest/TestComponentCSharp_Tests.net5.cs:29]: Would this call alone without try catch just be fine?  As technically any exceptions should fail the test right?

File: cswinrt/strings/additions/Windows.Storage.Streams/IBufferByteAccess.cs [hunk 3]
```diff
@@ -1,7 +1,8 @@
 namespace Windows.Storage.Streams
 {
-    using global::System;
-
+    using global::System;
+
+    [global::WinRT.WindowsRuntimeType("Windows.Foundation.UniversalApiContract")]
```
Comment (manodasanW) [inline cswinrt/strings/additions/Windows.Storage.Streams/IBufferByteAccess.cs:5]: Is this a class issue that we need to look at our other additions and add it to them also if they are used in similar ways?  Or was this a one-off.

File: UnitTest/TestComponentCSharp_Tests.net5.cs (hunk 1 above)
Comment (j0shuams) [reply UnitTest/TestComponentCSharp_Tests.net5.cs:9]: OpenStreamForReadAsync is a net5 function, it isn't available in net2

File: UnitTest/TestComponentCSharp_Tests.net5.cs (hunk 2 above)
Comment (j0shuams) [reply UnitTest/TestComponentCSharp_Tests.net5.cs:29]: Wasn't sure about this, but I'll take away the assert trues and the try/catch

Comment (manodasanW) [reply UnitTest/TestComponentCSharp_Tests.net5.cs:29]: You can confirm by just commenting out your fix and then running the test.  It should fail on exception.

**Review by manodasanW — APPROVED** [review]

**Comment by j0shuams** [comment]
> I wrote some test for this:
> 
> ```cs
> using System;
> using System.IO;
> using System.Runtime.InteropServices.WindowsRuntime;
> using System.Threading.Tasks;
> using Windows.Storage.Streams;
> using Xunit;
> 
> namespace UnitTest
> {
>     public class StreamTests
>     {
>         [Fact]
>         public async Task WriteBuffer()
>         {
>             var random = new Random(42);
>             byte[] data = new byte[256];
>             random.NextBytes(data);
> 
>             using var stream = new InMemoryRandomAccessStream();
>             IBuffer buffer = data.AsBuffer();
>             await stream.WriteAsync(buffer);
>         }
> 
>         [Fact]
>         public async Task WriteAndRead()
>         {
>             var random = new Random(42);
>             byte[] data = new byte[256];
>             random.NextBytes(data);
> 
>             using var stream = new InMemoryRandomAccessStream().AsStream();
>             await stream.WriteAsync(data, 0, data.Length);
>             stream.Seek(0, SeekOrigin.Begin);
> 
>             byte[] read = new byte[256];
>             await stream.ReadAsync(read, 0, read.Length);
>             Assert.Equal(read, data);
>         }
>     }
> }
> ```
> 
> Could you confirm that these tests also get fixed?

Yes those tests pass now and have been added as unit tests. I also added a unit test based on issue 359, it should be fixed as well now.

**Pull request merged**